Go off the rails

gibbz00 opened the pull request:
No description required.

Depends on #31

gibbz00 committed:
refactor(ari): extract error to separate crate

gibbz00 committed:
refactor(ari): extract `Event` to separate crate

gibbz00 committed:
feat(ari): temporarily remove `#[instrument]` annotations

gibbz00 committed:
reactor!(ari): use top module star imports

gibbz00 committed:
reactor(ari): use idiomatic code ordering in `client.rs`

gibbz00 committed:
feat(ari): remove tracing events from client builder

gibbz00 committed:
refactor!(ari): encapsulate `Client` fields

gibbz00 committed:
feat(ari): use unbounded event channels

gibbz00 committed:
feat!(ari): inline websocket channel setup in client

gibbz00 committed:
refactor(ari): move `RequestClient` to its own file

gibbz00 committed:
refactor!(ari): invert ownership if `Channel` requests

gibbz00 committed:
refactor!(ari/channel): require `channel_id` only in request calls

gibbz00 committed:
refactor!(ari/channel): make fields private and supply getters instead

gibbz00 committed:
refactor!(ari/channel): initial `serde_qs` integration

gibbz00 committed:
refactor!(ari/channel): continued `serde_qs` integration

gibbz00 committed:
feat!(ari/channel): remove tracing event logging

Libraries should avoid locking in logging solutions. Future work could
instead instrument the function behind a feature flag.

gibbz00 committed:
perf(ari/channel): avoid recreating `reqwest::Client` on every request

gibbz00 committed:
refactor(ari): add `RequestClient.authorized_delete`

gibbz00 committed:
refactor(ari): add `RequestClient.authorized_post`

gibbz00 committed:
refactor(ari): add `RequestClient.authorized_post_json_response`

gibbz00 committed:
refactor(ari): add `RequestClient.authorized_post_variables`

gibbz00 committed:
refactor(ari): add `RequestClient.authorized_get`

gibbz00 committed:
refactor!(ari): make `DeviceStateChanged` private in `event` module

gibbz00 committed:
refactor!(ari): invert self to `RequestClient` in `playback`

seffradev commented on arirs/src/channel.rs:
```diff
@@ -34,822 +61,431 @@ pub enum OriginateParams<'a> {
-    Failure,
-    #[display("answered_elsewhere")]
-    AnsweredElsewhere,
+pub use reason::Reason;
+mod reason {
```
question: why?

seffradev commented on arirs/src/request_client/mod.rs:
question: why is this named `request_client/mod.rs` rather than `request_client.rs`?

seffradev commented on arirs/src/request_client/mod.rs:
```diff
@@ -0,0 +1,120 @@
+pub use core::RequestClient;
+mod core {
```
question: why?

seffradev commented on arirs/src/request_client/mod.rs:
```diff
@@ -0,0 +1,120 @@
+        }
+
+        pub(crate) async fn authorized_post_variables<T: Serialize, R: DeserializeOwned>(
+            &self,
+            path: impl AsRef<[&str]>,
+            params: T,
+            variables: &HashMap<&str, &str>,
+        ) -> Result<R> {
+            let url = self.authorized_url(path, params)?;
+            let response = self
+                .as_ref()
+                .post(url)
+                .json(&serde_json::json!({
+                    "variables": variables
+                }))
+                .send()
+                .await?
+                .json()
+                .await?;
+            Ok(response)
+        }
+
+        pub(crate) async fn authorized_delete<T: Serialize>(&self, path: impl AsRef<[&str]>, params: T) -> Result<()> {
+            let url = self.authorized_url(path, params)?;
+            self.as_ref().delete(url).send().await?;
+            Ok(())
+        }
+
+        fn authorized_url<'a, T: Serialize>(&self, path: impl AsRef<[&'a str]>, params: T) -> Result<Url> {
+            let mut url = self.url().join(&path.as_ref().join("/"))?;
+            self.set_authorized_query_params(&mut url, params);
+            Ok(url)
+        }
+
+        pub(crate) fn authorize_request<T>(&self, inner: T) -> AuthorizedRequest<T> {
+            AuthorizedRequest {
+                api_key: self.get_api_key(),
+                inner,
+            }
+        }
```
question: why are all these named something with `authorize`? I may be wrong, but I don't know of any unauthenticated operations in ARI.

gibbz00 replied:
fair :) I'll take a look at this once I've gained some more experience for working with the lib and ARI

seffradev commented on arirs/src/client.rs:
```diff
@@ -2,208 +2,107 @@ use std::time::Duration;
-                            event!(Level::INFO, "WebSocket closed");
-                            break;
+                            None => {
+                                tracing::error!("WebSocket closed");
```
question: is a websocket closing an error?

gibbz00 replied:
Probably, since we're the ones driving it. I would much rather return an error here. I'll be going over logging soon.

seffradev commented on arirs/src/channel.rs:
```diff
@@ -34,822 +61,431 @@ pub enum OriginateParams<'a> {
-    #[instrument(level = "debug")]
-    pub fn start_moh(&self, _client: &Client) -> Result<()> {
-        unimplemented!()
+impl RequestClient {
```
suggestion: if we're doing a refactor of how the different modules work, can't we abstract out the different model functionalities (`bridge`, `channel`, `playback`) to be implemented traits so we don't populate `RequestClient`s internal namespace?

gibbz00 replied:
Possibly, I was thinking about simply prefixing the methods with `playback_`, `bridge_` etc.

seffradev replied:
I still think that bloats the namespace a bit too much. It does improve the scope of functions but I generally liked the previous interface of `Channel::func(&client)` so I think a `use ChannelExt; client.originate(params);` would be cleaner. I also think this would improve the ability to e.g. mock which could be nice for tests. Though mocks are generally bad :)

gibbz00 replied:
I've also thought about implementing traits on the client. Problem is that a lot of functions already collide without a prefix (bridge/channel, live/stored recording, listing and getting by id). Prefixes will in many cases be needed anyways, or there will be quite a few cases where disambiguation is required, ex. `<RequestClient as ChannelClient>::stop_moh(&client, params)`, which I think should be avoided as much as possible.

There's probably also plenty of work that can be done do reduce the number of `ResponseClient` methods in the first place. Say by ombining the `_with_id()` methods to their counterparts by setting the id optional. Or say by using generics on the common create, list, get, remove/destroy metods. I think that this should be the primary focus point. The type system should be able to infer a whole lot if we typeset the ids and implement some trait with associated consts and types on them.

seffradev reviewed (changes requested):
Just leaving a partial review because I am having a lazy Saturday :)

gibbz00 committed:
refactor!(ari): cleanup and place `RtpStastistics` in `channel` module

gibbz00 committed:
refactor!(ari): invert self to `RequestClient` in `bridge`

gibbz00 committed:
feat!(ari): prefix `live_recording_` to live recording requests

gibbz00 committed:
feat!(ari): prefix `stored_recording_` to stored recording requests

gibbz00 committed:
refactor!(ari): invert self to `RequestClient` in `recording`

gibbz00 committed:
refactor!(ari): cleanup and place `Variable` in `channel` module

gibbz00 committed:
feat!(ari): go over struct derives and place them in separate files

- Place `Debug` first
- Request params derives `Serialize` and `Debug` only. Fields are
  all correctly serialized as "camelCase". (Rename all is sometimes
  superflously applied to avoid future compatibility issues.)
- Response data derives `Deserialize` only. Their fields are made
  private, exposed through `Getters` derive.

gibbz00 committed:
refactor(ari): place `channel` handling under `request_client` module

gibbz00 committed:
refactor(ari): place `channel` handling under `request_client` module

gibbz00 committed:
refactor(ari): place `bridge` and `channel` handling under `request_client` module

gibbz00 committed:
refactor(ari): place `recording` handling under `request_client` module

gibbz00 committed:
chore(ari): go over `RequestClient` method visibilities

gibbz00 committed:
perf(ari): memoize API key

gibbz00 committed:
chore(ari): make `AuthorizedRequest` private

gibbz00 committed:
refactor(ari): prefix bridge client requests with `bridge_`

gibbz00 committed:
refactor(ari): prefix channel client requests with `channel_`

gibbz00 committed:
refactor(ari): prefix playback client requests with `playback_`

gibbz00 committed:
feat!(ari): go over WebSocket event handling

Also removes `tracing` library lock-in.

gibbz00 commented:
> Just leaving a partial review because I am having a lazy Saturday :)

My fault that this is 40 commits long 😅

gibbz00 committed:
feat!(ari): add `ClientError`

Begin moving away from crate-wide error.

gibbz00 committed:
chore!(ari): remove `AriError::Tungstenite`

gibbz00 committed:
chore!(ari): remove `AriError::Unknown`

gibbz00 committed:
chore!(ari): remove `AriError::UrlParse`

gibbz00 committed:
refactor!(ari): replace `AriError` with `RequestClientError`

gibbz00 committed:
refactor(ari): remove fully qualified `std::result::Result` usage

gibbz00 committed:
refactor!(ari): rename `Client` to `Asterisk`

gibbz00 committed:
refactor!(ari): rename `RequestClient` to `AriClient`

gibbz00 committed:
chore(ari): go over tokio feature flags

gibbz00 committed:
chore(ari): replace qualified `chrono::Utc` with use

gibbz00 committed:
chore(ari): remove misleading patch version from `Cargo.toml`

gibbz00 commented:
Satisfied now :)
